File: app/src/main/java/com/budgetbuddy/data/repository/BadgeRepository.kt
package com.budgetbuddy.data.repository

import android.app.NotificationManager
import android.content.Context
import android.util.Log
import androidx.core.app.NotificationCompat
import com.budgetbuddy.BudgetBuddyApp
import com.budgetbuddy.R
import com.budgetbuddy.data.local.dao.BadgeDao
import com.budgetbuddy.data.local.dao.BudgetDao
import com.budgetbuddy.data.local.dao.DebtDao
import com.budgetbuddy.data.local.dao.GoalDao
import com.budgetbuddy.data.local.dao.TransactionDao
import com.budgetbuddy.data.local.entities.BadgeEntity
import com.budgetbuddy.data.local.entities.BadgeType
import com.budgetbuddy.util.DateUtils
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import java.util.Calendar
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BadgeRepository @Inject constructor(
    @ApplicationContext private val context: Context,
    private val badgeDao: BadgeDao,
    private val transactionDao: TransactionDao,
    private val budgetDao: BudgetDao,
    private val goalDao: GoalDao,
    private val debtDao: DebtDao
) {
    fun getBadges(userId: String): Flow<List<BadgeEntity>> = badgeDao.getBadges(userId)

    private suspend fun awardBadge(userId: String, type: BadgeType) {
        // Only award (and notify) if this badge hasn't been earned before
        if (badgeDao.getBadgeByType(userId, type) == null) {
            badgeDao.insertBadge(BadgeEntity(userId = userId, badgeType = type))
            showBadgeNotification(type)
            Log.i("BadgeRepository", "Badge awarded: ${type.name}")
        }
    }

    /**
     * Posts a heads-up notification congratulating the user on earning a new badge.
     * Each badge type uses a fixed notification ID so a second (impossible) award
     * would just update the existing notification rather than stacking duplicates.
     */
    private fun showBadgeNotification(type: BadgeType) {
        val (icon, name, message) = badgeNotificationContent(type)
        val notification = NotificationCompat.Builder(context, BudgetBuddyApp.CHANNEL_BADGE_ACHIEVEMENTS)
            .setSmallIcon(R.drawable.ic_launcher_foreground)
            .setContentTitle("$icon  Badge Unlocked — $name")
            .setContentText(message)
            .setStyle(NotificationCompat.BigTextStyle().bigText(message))
            .setPriority(NotificationCompat.PRIORITY_HIGH)
            .setAutoCancel(true)
            .build()

        val nm = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        // Use the badge type's ordinal as a stable, unique notification ID
        nm.notify(BADGE_NOTIF_BASE_ID + type.ordinal, notification)
    }

    private fun badgeNotificationContent(type: BadgeType): Triple<String, String, String> = when (type) {
        BadgeType.FIRST_STEP    -> Triple("🎯", "First Step",
            "You logged your first transaction. Your financial journey has begun!")
        BadgeType.DAILY_TRACKER -> Triple("🔥", "Daily Tracker",
            "7-day logging streak! Consistency is the key to financial control.")
        BadgeType.PERFECT_MONTH -> Triple("📅", "Perfect Month",
            "You earned more than you spent this month. Great work staying in the green!")
        BadgeType.BUDGET_MASTER -> Triple("💰", "Budget Master",
            "Every budget category stayed under its limit this month. Impressive discipline!")
        BadgeType.DEBT_SLAYER   -> Triple("⚔️", "Debt Slayer",
            "You paid off a debt in full. One less thing weighing on your finances!")
        BadgeType.GOAL_GETTER   -> Triple("🏆", "Goal Getter",
            "Savings goal completed! Set your next target and keep the momentum going.")
        BadgeType.THRIFTY_CHAMP -> Triple("💎", "Thrifty Champ",
            "You saved over 33% of your income this month. That's serious financial discipline!")
    }

    suspend fun checkAndAwardBadges(userId: String) {
        checkFirstStep(userId)
        checkLoggingStreak(userId)
        checkPerfectMonth(userId)
        checkBudgetMaster(userId)
        checkDebtSlayer(userId)
        checkGoalGetter(userId)
        checkThriftyChamp(userId)
    }

    /** Awarded for logging the very first transaction. */
    private suspend fun checkFirstStep(userId: String) {
        val count = transactionDao.getTransactionCountForDay(userId, 0L, System.currentTimeMillis())
        if (count >= 1) awardBadge(userId, BadgeType.FIRST_STEP)
    }

    /** Awarded for logging at least one transaction every day for 7 consecutive days. */
    private suspend fun checkLoggingStreak(userId: String) {
        var streakDays = 0
        for (i in 0 until 7) {
            val dayMs = System.currentTimeMillis() - (i * 86_400_000L)
            val count = transactionDao.getTransactionCountForDay(
                userId, DateUtils.startOfDay(dayMs), DateUtils.endOfDay(dayMs)
            )
            if (count > 0) streakDays++ else break
        }
        if (streakDays >= 7) awardBadge(userId, BadgeType.DAILY_TRACKER)
    }

    /** Awarded when total income exceeds total expenses in the current month. */
    private suspend fun checkPerfectMonth(userId: String) {
        val start = DateUtils.startOfMonth()
        val end = DateUtils.endOfMonth()
        val income = transactionDao.getTotalIncomeForPeriod(userId, start, end) ?: 0.0
        val expense = transactionDao.getTotalExpenseForPeriod(userId, start, end) ?: 0.0
        if (income > 0 && income > expense) awardBadge(userId, BadgeType.PERFECT_MONTH)
    }

    /** Awarded when no budget category is exceeded (spent < limit) in the current month. */
    private suspend fun checkBudgetMaster(userId: String) {
        val cal = Calendar.getInstance()
        val month = cal.get(Calendar.MONTH) + 1  // BudgetEntity stores 1-based month
        val year = cal.get(Calendar.YEAR)
        val start = DateUtils.startOfMonth()
        val end = DateUtils.endOfMonth()
        val budgets = budgetDao.getBudgetsForMonthOnce(userId, month, year)
        if (budgets.isEmpty()) return  // no budgets set — badge not yet earnable
        val allUnderLimit = budgets.all { budget ->
            val spent = transactionDao.getTotalExpenseByCategoryAndPeriod(
                userId, budget.categoryId, start, end
            ) ?: 0.0
            spent < budget.limitAmount
        }
        if (allUnderLimit) awardBadge(userId, BadgeType.BUDGET_MASTER)
    }

    /** Awarded when the user fully pays off at least one debt. */
    private suspend fun checkDebtSlayer(userId: String) {
        if (debtDao.getPaidOffDebtsCount(userId) >= 1) awardBadge(userId, BadgeType.DEBT_SLAYER)
    }

    /** Awarded when the user completes at least one savings goal. */
    private suspend fun checkGoalGetter(userId: String) {
        if (goalDao.getCompletedGoalsCount(userId) >= 1) awardBadge(userId, BadgeType.GOAL_GETTER)
    }

    /** Awarded when income is at least 1.5× expenses this month (saved ≥ 33% of income). */
    private suspend fun checkThriftyChamp(userId: String) {
        val start = DateUtils.startOfMonth()
        val end = DateUtils.endOfMonth()
        val income = transactionDao.getTotalIncomeForPeriod(userId, start, end) ?: 0.0
        val expense = transactionDao.getTotalExpenseForPeriod(userId, start, end) ?: 0.0
        if (income > 0 && income >= expense * 1.5) awardBadge(userId, BadgeType.THRIFTY_CHAMP)
    }

    companion object {
        // Notification IDs for badges start at 2000 to avoid clashing with budget (1000+) notifications
        private const val BADGE_NOTIF_BASE_ID = 2000
    }
}

File: ios/BudgetBuddy/Repositories/BadgeRepository.swift
import Foundation
import SwiftData
import UserNotifications

final class BadgeRepository {
    private let context: ModelContext
    private let firestore: FirestoreRepository
    private let transactionRepo: TransactionRepository
    private let budgetRepo: BudgetRepository
    private let goalRepo: GoalRepository
    private let debtRepo: DebtRepository

    init(context: ModelContext, firestore: FirestoreRepository,
         transactionRepo: TransactionRepository, budgetRepo: BudgetRepository,
         goalRepo: GoalRepository, debtRepo: DebtRepository) {
        self.context         = context
        self.firestore       = firestore
        self.transactionRepo = transactionRepo
        self.budgetRepo      = budgetRepo
        self.goalRepo        = goalRepo
        self.debtRepo        = debtRepo
    }

    func all(userId: String) throws -> [BBBadge] {
        let pred = #Predicate<BBBadge> { $0.userId == userId }
        return try context.fetch(FetchDescriptor<BBBadge>(predicate: pred,
                                                          sortBy: [SortDescriptor(\.earnedAt, order: .reverse)]))
    }

    func checkAndAwardBadges(userId: String) async {
        await checkFirstStep(userId: userId)
        await checkLoggingStreak(userId: userId)
        await checkPerfectMonth(userId: userId)
        await checkBudgetMaster(userId: userId)
        await checkDebtSlayer(userId: userId)
        await checkGoalGetter(userId: userId)
        await checkThriftyChamp(userId: userId)
    }

    // MARK: - Private checks

    private func awardBadge(userId: String, type: BadgeType) async {
        let pred = #Predicate<BBBadge> { $0.userId == userId && $0.badgeType == type }
        guard (try? context.fetchCount(FetchDescriptor<BBBadge>(predicate: pred))) == 0 else { return }
        let badge = BBBadge(userId: userId, badgeType: type)
        context.insert(badge)
        try? context.save()
        Task { await firestore.saveBadge(userId: userId, badge: badge) }
        postNotification(type: type)
    }

    private func checkFirstStep(userId: String) async {
        let count = (try? transactionRepo.countInRange(userId: userId,
                                                       start: .distantPast, end: .now)) ?? 0
        if count >= 1 { await awardBadge(userId: userId, type: .firstStep) }
    }

    private func checkLoggingStreak(userId: String) async {
        var streak = 0
        for i in 0..<7 {
            let day   = Calendar.current.date(byAdding: .day, value: -i, to: Date())!
            let start = DateUtils.startOfDay(day)
            let end   = DateUtils.endOfDay(day)
            let count = (try? transactionRepo.countInRange(userId: userId, start: start, end: end)) ?? 0
            if count > 0 { streak += 1 } else { break }
        }
        if streak >= 7 { await awardBadge(userId: userId, type: .dailyTracker) }
    }

    private func checkPerfectMonth(userId: String) async {
        let start  = DateUtils.startOfMonth()
        let end    = DateUtils.endOfMonth()
        let income = (try? transactionRepo.totalIncome(userId: userId, start: start, end: end)) ?? 0
        let spend  = (try? transactionRepo.totalExpense(userId: userId, start: start, end: end)) ?? 0
        if income > 0 && income > spend { await awardBadge(userId: userId, type: .perfectMonth) }
    }

    private func checkBudgetMaster(userId: String) async {
        let cal   = Calendar.current
        let month = cal.component(.month, from: Date())
        let year  = cal.component(.year, from: Date())
        let start = DateUtils.startOfMonth()
        let end   = DateUtils.endOfMonth()
        guard let budgets = try? budgetRepo.forMonth(userId: userId, month: month, year: year),
              !budgets.isEmpty else { return }
        let allUnder = budgets.allSatisfy { budget in
            let spent = (try? transactionRepo.totalExpenseByCategory(
                userId: userId, categoryId: budget.categoryId, start: start, end: end)) ?? 0
            return spent < budget.limitAmount
        }
        if allUnder { await awardBadge(userId: userId, type: .budgetMaster) }
    }

    private func checkDebtSlayer(userId: String) async {
        let count = (try? debtRepo.paidOffCount(userId: userId)) ?? 0
        if count >= 1 { await awardBadge(userId: userId, type: .debtSlayer) }
    }

    private func checkGoalGetter(userId: String) async {
        let count = (try? goalRepo.completedCount(userId: userId)) ?? 0
        if count >= 1 { await awardBadge(userId: userId, type: .goalGetter) }
    }

    private func checkThriftyChamp(userId: String) async {
        let start  = DateUtils.startOfMonth()
        let end    = DateUtils.endOfMonth()
        let income = (try? transactionRepo.totalIncome(userId: userId, start: start, end: end)) ?? 0
        let spend  = (try? transactionRepo.totalExpense(userId: userId, start: start, end: end)) ?? 0
        if income > 0 && income >= spend * 1.5 { await awardBadge(userId: userId, type: .thriftyChamp) }
    }

    private func postNotification(type: BadgeType) {
        let content = UNMutableNotificationContent()
        content.title   = "\(type.icon)  Badge Unlocked — \(type.displayName)"
        content.body    = type.description
        content.sound   = .default
        let req = UNNotificationRequest(identifier: "badge_\(type.rawValue)",
                                        content: content, trigger: nil)
        UNUserNotificationCenter.current().add(req)
    }
}
